zramSwap: allow configure compression algorithm + cleanups

danbst opened the pull request:
###### Motivation for this change
As I mentioned in https://discourse.nixos.org/t/install-nixos-on-512-mb-ram/1713, I'm using zstd compression, but current module doesn't allow to change compressor declaratively. Here we go, and fix few problems with this module.

###### Things done

- add `zramSwap.algorithm` option, which allows to change compressor
declaratively
- some documentation changes
- replaced `/sys/block/zram*` handling with `zramctl`, because I had occasional
"Device is busy" error (looks like zram has to be configured in predefined order)
- added `memoryPercent` and `algorithm` as restart triggers. I think, it was
a bug that changing `memoryPercent` in configuration wasn't applied immediately.
- removed a bind to .swap device. While it looks natural (when swap device goes
off, so should zram device), it wasn't implemented properly. This caused problems
with swapon/swapoff:
```
$ cat /proc/swaps
Filename                                Type            Size    Used    Priority
/dev/zram0                              partition       8166024 0       -2
/var/swapfile                           file            5119996 5120    1

$ sudo swapoff -a

$ sudo swapon -a
swapon: /dev/zram0: read swap header failed

$ cat /proc/swaps
Filename                                Type            Size    Used    Priority
/var/swapfile                           file            5119996 0       1
```

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @Mic92

danbst committed:
zramSwap: allow configure compression algorithm + cleanups

- add `zramSwap.algorithm` option, which allows to change compressor
declaratively. zstd as default
- add `zramSwap.swapDevices` option, which allows to define how many zram
devices will be used as swap. Rest devices can be managed freely
- simpler floating calculations
- fix udev race condition
- some documentation changes
- replaced `/sys/block/zram*` handling with `zramctl`, because I had occasional
"Device is busy" error (looks like zram has to be configured in predefined order)
- added `memoryPercent` and `algorithm` as restart triggers. I think, it was
a bug that changing `memoryPercent` in configuration wasn't applied immediately.
- removed a bind to .swap device. While it looks natural (when swap device goes
off, so should zram device), it wasn't implemented properly. This caused problems
with swapon/swapoff:
```
$ cat /proc/swaps
Filename                                Type            Size    Used    Priority
/dev/zram0                              partition       8166024 0       -2
/var/swapfile                           file            5119996 5120    1

$ sudo swapoff -a

$ sudo swapon -a
swapon: /dev/zram0: read swap header failed

$ cat /proc/swaps
Filename                                Type            Size    Used    Priority
/var/swapfile                           file            5119996 0       1
```

danbst commented:
@worldofpeace yeah, according to gist https://gist.github.com/asssaf/f9bef956a755e67ed742e3265765e48d it is possible to do only with udev. However I'm not sure udev solution will apply required changes when options are changed. For example, when you change compressor and rebuild system, does it rebuild zram devices with new algo?

danbst commented:
Alright, I've fixed things up.
- one more old race condition hunted!
- replaced integer math with floating math!
- introduced `swapDevices`. In general, zram and swap are not same, you can allocate zram device and use it as regular block device, including compressed tmpfs usage. So, one can set 1 device for swap, and 1 for other things. Previously it required lots of hacking.
- for those who use several zram devices added a warning that this may be misusage. The warning can be disabled by explicitly setting `zramSwap.swapDevices`.
- documentation, documentation
- and finally, **zstd is default now**!

cc for review @worldofpeace @Infinisil @bjornfor @wmertens @wizeman

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/install-nixos-on-512-mb-ram/1713/13

Mic92 commented on nixos/modules/config/zram.nix:
```diff
@@ -58,12 +91,26 @@ in
+        description = ''
+          Compression algorithm. <literal>lzo</literal> has good compression,
+          but is slow. <literal>lz4</literal> has bad compression, but is fast.
+          <literal>zstd</literal> is both good compression and fast (kernel 4.18+).
```
There could be a warning generated if the kernel is too old (config.boot.kernelPackages.kernel.version)

danbst replied:
I'd better remove this section. Current kernel is already capable of zstd zram.

Mic92 commented on nixos/modules/config/zram.nix:
```diff
@@ -96,29 +142,34 @@ in
-              totalmem=$(${pkgs.gnugrep}/bin/grep 'MemTotal: ' /proc/meminfo | ${pkgs.gawk}/bin/awk '{print $2}')
-              mem=$(((totalmem * ${toString cfg.memoryPercent} / 100 / ${toString cfg.numDevices}) * 1024))
+              totalmem=$(${pkgs.gawk}/bin/awk '/MemTotal: /{print $2}' /proc/meminfo)
+              mem=$(echo "$totalmem*${toString cfg.memoryPercent}/100.0/${toString devicesCount}*1024" | ${bc})
```
Do we depend on `bc` in the minimal closure?

Mic92 replied:
Otherwise gawk could also perform floating point calculation I think.

danbst commented:
@Mic92 addressed your issues. Also, added some release notes.

I'll merge in a week if nobody beats me

danbst committed:
zramSwap: remove basic.target for zram devices

This creates a dependency cycle when used with boot.tmpOnTmpfs:
basic.target <- tmp.mount <- swap.target <- zram-init-dev0 <- basic.target

This same fix is done already for tmp.mount

Fixes https://github.com/NixOS/nixpkgs/issues/47474

xaverdh commented:
I can confirm, that the patch resolves the dependency cycle.
